src/configuration/prima.rs

```diff
@@ -97,4 +109,14 @@ mod tests {
         //assert_eq!(None, "".parse::<Environment>().err());
         assert_eq!(None, "whatever".parse::<Environment>().ok());
     }
+
+    #[test]
+    pub fn test_tags() {
+        let count = 1;
+        let config = PrimaConfiguration::new("to_addr", "from_addr", "namespace", Environment::Dev)
+            .with_tag("key", &"value")
+            .with_tag("count", &count);
```

Comment:
Does the API constrain users to pass the second parameter as a reference? Or is it just the test?

Reply:
Just played a bit with it in the Rust playground, you can avoid forcing the user to pass a reference by accepting a `T` instead of `&T` in with_tag signature.
The user can still pass a reference if they want, since in that case `T` will be replaced with a reference type.

Reply:
To me, the semantic meaning of "val: T" is "this fn takes ownership of the T you give it", the semantic meaning of "val: &T" is "this fn needs read access to the T you give it" - when I read an interface that requires a T, I assume it means I have to move the T - having to know that `impl<T> Display for &T where T: Display` exists in order to know the interface lets me pass a reference is a lot less obvious than just "oh this takes a reference, I can just give it a reference".

I wouldn't use the term "forcing the user to pass a reference" - a reference is strictly less constrained for a caller to pass than the value itself.

Reply:
Before addressing your point, let me explain why I'm suggesting to change the signature.
As it is now, `with_tag` must always be called with a reference, even if the second parameter is a plain number or a string literal.
Why should I have to write
`with_tag("key", &3)`
`with_tag("key", &"aaa")`

instead of

`with_tag("key", 3)`
`with_tag("key", "aaa")`?

It is a constraint that looks very weird and innatural to me.

Now we can get to the technical reasoning. The premise

> the semantic meaning of "val: T" is "this fn takes ownership of the T you give it"

is not correct, because T can get replaced with whatever sized type you throw at it, including reference types.
If `with_tag` accepts a `T: ToString` and you have a variable `a` of type `String`, calling `with_tag("key", &a)` works just fine because `T` gets replaced with `&String`.

In the end it's just a matter of keeping the signature as loose as possible to make life easier to callers.

Reply:
>Now we can get to the technical reasoning. The premise
> the semantic meaning of "val: T" is "this fn takes ownership of the T you give it"
>is not correct, because T can get replaced with whatever sized type you throw at it, including reference types.

Yes, at which point the fn /takes ownership/ of the reference type - the semantic meaning is the same regardless of the type of T, and the fact that T /can/ be a reference here is only because `impl<T> Display for &T where T: Display` exists. I'm not saying T can't be a reference type, I'm saying that the meaning of an interface that takes `val: T` is "this fn takes ownership of the T passed".

For a user to know they can pass a reference if the arg is a raw `T`, they have to know that the particular `impl Display` exists, instead of just reading the type signature - it's just another thing to hold in their head. This is made worse by the fact ide's won't suggest "hey because of such and such trait impl you can make this a reference" if they want use the passed value later - they just get a "use of moved value" error, e.g:
```
let count = String::from("test");
let config = PrimaConfiguration::new("to_addr", "from_addr", "namespace", Environment::Dev)
    .with_tag("key", &"value")
    .with_tag("count", count);
println!("{}", count);

// Error:
error[E0382]: borrow of moved value: `count`
   --> src/configuration/prima.rs:115:24
    |
111 |         let count = String::from("test");
    |             ----- move occurs because `count` has type `String`, which does not implement the `Copy` trait
...
114 |             .with_tag("count", count);
    |                                ----- value moved here
115 |         println!("{}", count);
    |                        ^^^^^ value borrowed here after move
    |
```

Reply:
> This is made worse by the fact ide's won't suggest "hey because of such and such trait impl you can make this a reference" if they want use the passed value later - they just get a "use of moved value" error

I get this point. It's not that obvious that you can pass in a reference in, which may lead to unnecessary cloning as @WilliamVenner said. It could be an acceptable trade-off.

I'm still not convinced  by these two claims:
> the fact that T /can/ be a reference here is only because impl<T> Display for &T where T: Display exists

> the meaning of an interface that takes val: T is "this fn takes ownership of the T passed"

But that will be an interesting technical discussion for another time.
I'm definitely not going to block the PR for this! 🙌

Reply:
Cheers Francesco, maybe we should schedule some time to argue about this stuff eh :laughing: